ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+	resources, err := client.getResources()
+	if err != nil {
+		return nil, err
+	}
+	resourceMap := make(map[string]*ecsmodel.Resource)
+	for _, resource := range resources {
+		if *resource.Name == "PORTS" {
+			// Except for RCI, TCP Ports are named as PORTS_TCP in Agent for Host Resources purpose.
+			resource.Name = aws.String("PORTS_TCP")
+		}
+		resourceMap[*resource.Name] = resource
+	}
+	return resourceMap, nil
+}
+
+func getCpuAndMemory() (int64, int64) {
+	memInfo, err := system.ReadMemInfo()
+	mem := int64(0)
+	if err == nil {
+		mem = memInfo.MemTotal / 1024 / 1024 // MiB
+	} else {
+		logger.Error("Unable to get memory info", logger.Fields{
+			field.Error: err,
+		})
+	}
+
+	cpu := runtime.NumCPU() * 1024
+
+	return int64(cpu), mem
+}
+
+func validateRegisteredAttributes(expectedAttributes, actualAttributes []*ecsmodel.Attribute) error {
+	var err error
+	expectedAttributesMap := attributesToMap(expectedAttributes)
+	actualAttributesMap := attributesToMap(actualAttributes)
+	missingAttributes, err := findMissingAttributes(expectedAttributesMap, actualAttributesMap)
+	if err != nil {
+		msg := strings.Join(missingAttributes, ",")
+		logger.Error("Error registering attributes", logger.Fields{
+			field.Error:         err,
+			"missingAttributes": msg,
+		})
+	}
+	return err
+}
+
+func (client *ecsClient) getAdditionalAttributes() []*ecsmodel.Attribute {
+	attrs := []*ecsmodel.Attribute{
+		{
+			Name:  aws.String(osTypeAttrName),
+			Value: aws.String(client.configAccessor.OSType()),
+		},
+		{
+			Name:  aws.String(osFamilyAttrName),
+			Value: aws.String(client.configAccessor.OSFamily()),
+		},
+	}
+	// Send CPU arch attribute directly when running on external capacity. When running on EC2 or Fargate launch type,
+	// this is not needed since the CPU arch is reported via instance identity document in those cases.
+	if client.configAccessor.External() {
+		attrs = append(attrs, &ecsmodel.Attribute{
+			Name:  aws.String(cpuArchAttrName),
+			Value: aws.String(getCPUArch()),
+		})
+	}
+	return attrs
+}
+
+func (client *ecsClient) getOutpostAttribute(outpostARN string) []*ecsmodel.Attribute {
+	if len(outpostARN) > 0 {
+		return []*ecsmodel.Attribute{
+			{
+				Name:  aws.String("ecs.outpost-arn"),
+				Value: aws.String(outpostARN),
+			},
+		}
+	}
+	return []*ecsmodel.Attribute{}
+}
+
+func (client *ecsClient) getCustomAttributes() []*ecsmodel.Attribute {
+	var attributes []*ecsmodel.Attribute
+	for attribute, value := range client.configAccessor.InstanceAttributes() {
+		attributes = append(attributes, &ecsmodel.Attribute{
+			Name:  aws.String(attribute),
+			Value: aws.String(value),
+		})
+	}
+	return attributes
+}
+
+func (client *ecsClient) SubmitTaskStateChange(change ecs.TaskStateChange) error {
+	if change.Attachment != nil && client.stscAttachmentCustomRetryBackoff != nil {
+		retryFunc := func() error {
+			err := client.submitTaskStateChange(change)
+			if err == nil {
```

Comment:
nit: we seem to prefer
```
if err != nil {
 // Code
}
return nil
```
Over
```
if err == nil {
  return nil
}
// Code
```

Thoughts?

Reply:
Synced offline, this is not blocking and we will keep as is since it mirrors current existing code/behavior.